packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,384 @@
+        # Send the Envelope to the Agent's InBox.
+        assert self.in_queue is not None, "Input queue not initialized."
+        assert self.loop is not None, "Loop not initialized."
+        asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self.loop)
+
+    def _send_response(self):
+        """Pass the response from out-bounded Envelope back to the front-end client.
+
+        Currently, the response will be written back in the cmd terminal.
+        """
+        # self.wfile.write(self.path.encode())
+        pass
+
+    def connect(self):
+        """
+        Connect.
+
+        Upon HTTP Channel connection, kickstart the HTTP Server.
+        """
+        with self.lock:
+            if self.stopped:
+                self.stopped = False
+                # self.thread.start()
+                # Initializing the internal HTTP server
+                try:
+                    # self.httpd = HTTPServer((self.host, self.port), HTTPHandler(self))
+                    # handler = SimpleHTTPServer.SimpleHTTPRequestHandler
+                    # handler = HTTPHandler.set_channel(self)
+                    # handler.set_channel(self)
+                    self.httpd = HTTPServer(
+                        (self.host, self.port), HTTPHandlerFactory(self)
+                    )
+                    # self.httpd = HTTPServer((self.host, self.port), handler)
+                    logger.info(f"HTTP Server has connected to port {self.port}.")
+                    self.httpd.serve_forever()
+                except OSError:
+                    logger.error(
+                        f"{self.host}:{self.port} is already in use, please try another Socket."
+                    )
+
+    def disconnect(self) -> None:
+        """
+        Disconnect.
+
+        Shut-off the HTTP Server and then HTTP Channel.
+        """
+        with self.lock:
+            if not self.stopped:
+                self.httpd.shutdown()
+                logger.info(f"HTTP Server has shutdown on port {self.port}.")
+                self.stopped = True
+                # self.thread.join()
+
+
+def HTTPHandlerFactory(channel: HTTPChannel):
+    class HTTPHandler(BaseHTTPRequestHandler):
+        def __init__(self, *args, **kwargs):
+            self._channel = channel
+            super(HTTPHandler, self).__init__(*args, **kwargs)
+
+        def do_HEAD(self):
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+        def do_GET(self):
+            """Respond to a GET request."""
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+            parsed_path = urlparse(self.path)
+            url = parsed_path.path
+            # url = parsed_path.geturl()
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
+            self._channel._process_request("GET", url, param)
```

Comment:
Note, since this is a get request, you'd need to block here and wait for a response from this call.

So something like `response = self._channel._process_request("GET", url, param)` and then serve the response: https://pymotw.com/2/BaseHTTPServer/

Reply:
I renamed it to `_process_incoming_request`. I first changed this function so that it returns an Envelope. Next I input this Envelope into verification helper object **APIVerificationHelper** (see below) that will then output the response.